### DISCUSSION TITLE
2D cylindrical mesh generation

### DISCUSSION THREAD

**Discussion by xyyrly** [pybamm-team/PyBaMM#3938] [Q&A]
Hi Team,

Is there any example for 2D cylindrical mesh generation in Pybamm? Thank u very much!

**Comment by valentinsulzer** [comment]
No. If / when we add cylindrical coordinates, it will only be for the axisymmetric case (assuming the variable only changes in the r direction, not theta or z).

**Comment by xyyrly** [comment]
Thank you for your reply~ I notice 2D mesh is used in a thermal model,  is there any example to show how to generate 2D cartesian mesh in Pybamm then?

**Reply by valentinsulzer** [reply]
There isn't a clean example for this. If the domain you want to be 2D cartesian is "my domain", here is a quick example
```python
geometry = {"my domain": {"y": {"min": 0, "max": 1}, "z": {"min": 0, "max": 1}}}
var_pts = {"y": 10, "z": 10}
submesh_types = {"my domain": pybamm.ScikitUniform2DSubMesh}
spatial_methods = {"my domain": pybamm.ScikitFiniteElement()}
```
then pass those 4 arguments to `Simulation` alongside your model. That should work but I haven't tested it

**Reply by xyyrly** [reply]
Thank u very much! The following is what I tried, but i am not sure how to set the boundary conditions in both y and z direction now...
model = pybamm.BaseModel()
c = pybamm.Variable("Concentration", domain="my domain")
N = -pybamm.grad(c)  # define the flux
dcdt = -pybamm.div(N)  # define the rhs equation
model.rhs[c] = dcdt
model.boundary_conditions[c] = {
    "left": (pybamm.Scalar(0), "Neumann"),
    "right": (pybamm.Scalar(0), "Neumann"),
    "left": (pybamm.Scalar(0), "Neumann"),
    "right": (pybamm.Scalar(0), "Neumann"),
}
model.initial_conditions[c] =pybamm.Scalar(14630)
model.variables = {"Concentration": c,}
geometry = {"my domain": {"y": {"min": 0, "max": 1}, "z": {"min": 0, "max": 1}}}
var_pts = {"y": 10, "z": 10}
submesh_types = {"my domain": pybamm.ScikitUniform2DSubMesh}
mesh = pybamm.Mesh(geometry, submesh_types, var_pts)
spatial_methods = {"my domain": pybamm.ScikitFiniteElement()}
disc = pybamm.Discretisation(mesh, spatial_methods)
disc.process_model(model);

**Reply by valentinsulzer** [reply]
Digging further into this it looks like there are some hard-coded assumptions about tabs so the code I suggested might not work. @rtimms knows this part of the code better

**Reply by rtimms** [reply]
Yes, at the moment you have to specify the locations of the positive and negative tab on the boundary of the 2D domain using the “tabs” key in the geometry you define (see [here](https://github.com/pybamm-team/PyBaMM/blob/6b5f5b7fbff076a5abb6619b574e8c2c181381c2/pybamm/geometry/battery_geometry.py#L112)). 

You can see how boundary conditions are applied on tabs [here](https://github.com/pybamm-team/PyBaMM/blob/6b5f5b7fbff076a5abb6619b574e8c2c181381c2/pybamm/models/submodels/current_collector/potential_pair.py#L110). The boundary conditions on the “non-tab” regions turn into source terms in the weak formulation, so become part of the algebraic equation (see [here](https://github.com/pybamm-team/PyBaMM/blob/6b5f5b7fbff076a5abb6619b574e8c2c181381c2/pybamm/models/submodels/thermal/pouch_cell/pouch_cell_2D_current_collectors.py#L78) for an example where the “non-tab” boundary conditions aren’t zero flux).

**Reply by xyyrly** [reply]
Thank u very mcuh!
